File: gallery/scrape.mjs
import { chromium } from "playwright";
import { writeFileSync, mkdirSync } from "fs";
import { readdir } from "fs/promises";
import path from "path";

const QUERY = process.argv[2] ?? "تم تعبئة الكرش";
const IMAGES_DIR = "./images";
const SCROLL_TIMES = 15;
const MIN_BYTES = 5_000;
const IMG_EXTS = new Set(["jpg", "jpeg", "png", "gif", "webp", "avif"]);
const SKIP_HOSTS = ["google.com", "gstatic.com", "googleapis.com", "googleusercontent.com"];

mkdirSync(IMAGES_DIR, { recursive: true });

const url = QUERY.startsWith("http")
  ? QUERY
  : `https://www.google.com/search?q=${encodeURIComponent(QUERY)}&udm=2`;

console.log("Scraping:", url);

const browser = await chromium.launch({
  headless: false,
  args: ["--no-sandbox", "--disable-blink-features=AutomationControlled"],
});

const ctx = await browser.newContext({
  userAgent:
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36",
  viewport: { width: 1440, height: 900 },
});

// intercept image responses directly from the network
const intercepted = new Set();
ctx.on("response", async (res) => {
  const ct = res.headers()["content-type"] ?? "";
  if (!ct.startsWith("image/")) return;
  const resUrl = res.url();
  if (SKIP_HOSTS.some((h) => resUrl.includes(h))) return;
  if (resUrl.startsWith("data:")) return;
  try {
    const body = await res.body();
    if (body.length < MIN_BYTES) return;
    intercepted.add({ url: resUrl, body, ct });
  } catch {}
});

const page = await ctx.newPage();
await page.goto(url, { waitUntil: "networkidle", timeout: 30_000 });
await page.waitForTimeout(3000);

// also parse "ou" URLs from page source
function extractFromSource(html) {
  const urls = new Set();
  for (const m of html.matchAll(/"ou"\s*:\s*"(https?:\/\/[^"]+)"/g)) {
    urls.add(m[1]);
  }
  for (const m of html.matchAll(/https?:\/\/[^\s"'<>]+\.(?:jpg|jpeg|png|webp|gif)/gi)) {
    const u = m[0].replace(/[\\,;)]+$/, "");
    const host = u.match(/https?:\/\/([^/]+)/)?.[1] ?? "";
    if (!SKIP_HOSTS.some((h) => host.endsWith(h))) urls.add(u);
  }
  return urls;
}

const sourceUrls = new Set();
for (let i = 0; i < SCROLL_TIMES; i++) {
  await page.keyboard.press("End");
  await page.waitForTimeout(1800);
  const html = await page.content();
  for (const u of extractFromSource(html)) sourceUrls.add(u);

  const btn = await page.$("input[value='Show more results']");
  if (btn) { await btn.click(); await page.waitForTimeout(1200); }

  console.log(`  scroll ${i + 1}/${SCROLL_TIMES}: ${intercepted.size} intercepted, ${sourceUrls.size} from source`);
}

await browser.close();

// save intercepted images first
let saved = 0;
for (const { url: u, body, ct } of intercepted) {
  const ext = ct.split("/")[1]?.split(";")[0]?.trim().replace("jpeg", "jpg") ?? "jpg";
  const dest = path.join(IMAGES_DIR, `img_${String(saved).padStart(4, "0")}.${ext}`);
  writeFileSync(dest, body);
  console.log(`  [${++saved}] intercepted: ${dest} (${Math.round(body.length / 1024)}KB)`);
}

// download remaining source URLs
for (const u of sourceUrls) {
  try {
    const res = await fetch(u);
    const ct = res.headers.get("content-type") ?? "";
    if (!ct.startsWith("image/")) continue;
    const buf = Buffer.from(await res.arrayBuffer());
    if (buf.length < MIN_BYTES) continue;
    const ext = ct.split("/")[1]?.split(";")[0]?.trim().replace("jpeg", "jpg") ?? "jpg";
    const dest = path.join(IMAGES_DIR, `img_${String(saved).padStart(4, "0")}.${ext}`);
    writeFileSync(dest, buf);
    console.log(`  [${++saved}] fetched: ${dest} (${Math.round(buf.length / 1024)}KB)`);
  } catch {}
}

// build index.html
const files = (await readdir(IMAGES_DIR))
  .filter((f) => IMG_EXTS.has(f.split(".").pop()?.toLowerCase() ?? ""))
  .sort();

const imgTags = files.map((f) => `    <img src="images/${f}" loading="lazy">`).join("\n");

writeFileSync(
  "./index.html",
  `<!DOCTYPE html>
<html lang="ar" dir="rtl">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Gallery</title>
  <style>
    body { margin: 0; background: #111; }
    .grid {
      display: grid;
      grid-template-columns: repeat(auto-fill, minmax(240px, 1fr));
      gap: 4px;
      padding: 4px;
    }
    img { width: 100%; aspect-ratio: 1; object-fit: cover; display: block; }
  </style>
</head>
<body>
  <div class="grid">
${imgTags}
  </div>
</body>
</html>
`
);

console.log(`\nBuilt index.html with ${files.length} images`);

File: gallery/scrape.py
#!/usr/bin/env python3
"""
Scrape Google Images and build index.html gallery.
Usage: uv run scrape.py "your search query"
       uv run scrape.py "https://www.google.com/search?q=...&udm=2"
"""

import asyncio
import sys
import re
from pathlib import Path
from urllib.parse import quote

import httpx
from playwright.async_api import async_playwright

HERE = Path(__file__).parent
IMAGES_DIR = HERE / "images"
IMAGES_DIR.mkdir(exist_ok=True)

IMG_EXTS = {".jpg", ".jpeg", ".png", ".gif", ".webp", ".avif"}
SCROLL_TIMES = 15
MIN_SIZE = 5_000  # skip tiny images (icons, thumbnails < 5KB)


def make_url(query_or_url: str) -> str:
    if query_or_url.startswith("http"):
        return query_or_url
    return f"https://www.google.com/search?q={quote(query_or_url)}&udm=2"


def load_firefox_cookies():
    try:
        import browser_cookie3
        cj = browser_cookie3.firefox(domain_name=".google.com")
        return [
            {
                "name": c.name,
                "value": c.value,
                "domain": c.domain if c.domain else ".google.com",
                "path": c.path or "/",
            }
            for c in cj
        ]
    except Exception as e:
        print(f"No Firefox cookies: {e}")
        return []


async def collect_image_urls(page) -> set[str]:
    urls: set[str] = set()
    skip_domains = {"google.com", "gstatic.com", "googleapis.com", "googleusercontent.com"}

    for i in range(SCROLL_TIMES):
        await page.keyboard.press("End")
        await page.wait_for_timeout(2000)

        # extract full-size image URLs embedded in Google's JS data
        content = await page.content()

        # pattern 1: "ou":"URL" (original url field in JSON blobs)
        for m in re.finditer(r'"ou"\s*:\s*"(https?://[^"]+)"', content):
            urls.add(m.group(1))

        # pattern 2: standalone https:// image URLs not from google domains
        for m in re.finditer(r'https?://[^\s"\'<>]+\.(?:jpg|jpeg|png|webp|gif)', content, re.IGNORECASE):
            url = m.group(0).rstrip("\\,;)")
            host = re.search(r'https?://([^/]+)', url)
            if host and not any(d in host.group(1) for d in skip_domains):
                urls.add(url)

        # click "Show more results" if present
        btn = await page.query_selector("input[value='Show more results']")
        if btn:
            await btn.click()
            await page.wait_for_timeout(1500)

        print(f"  scroll {i+1}/{SCROLL_TIMES}: {len(urls)} urls found so far")

    return urls


async def download(urls: set[str]) -> int:
    saved = 0
    async with httpx.AsyncClient(follow_redirects=True, timeout=15) as client:
        for i, url in enumerate(sorted(urls)):
            try:
                r = await client.get(url)
                if len(r.content) < MIN_SIZE:
                    continue
                ct = r.headers.get("content-type", "image/jpeg")
                if "image" not in ct:
                    continue
                ext = ct.split("/")[-1].split(";")[0].strip().replace("jpeg", "jpg")
                if ext not in {"jpg", "png", "gif", "webp", "avif"}:
                    ext = "jpg"
                dest = IMAGES_DIR / f"img_{saved:04d}.{ext}"
                dest.write_bytes(r.content)
                saved += 1
                print(f"  [{saved}] saved {dest.name} ({len(r.content)//1024}KB)")
            except Exception as e:
                print(f"  [skip] {url[:70]}: {e}")
    return saved


async def main(query_or_url: str):
    url = make_url(query_or_url)
    print(f"Scraping: {url}")

    cookies = load_firefox_cookies()
    print(f"Loaded {len(cookies)} Firefox cookies")

    async with async_playwright() as p:
        browser = await p.firefox.launch(headless=False)
        ctx = await browser.new_context(
            user_agent=(
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:137.0) "
                "Gecko/20100101 Firefox/137.0"
            )
        )
        if cookies:
            await ctx.add_cookies(cookies)

        page = await ctx.new_page()
        await page.goto(url, wait_until="networkidle", timeout=30000)
        await page.wait_for_timeout(3000)

        urls = await collect_image_urls(page)
        await browser.close()

    print(f"\nFound {len(urls)} candidate image URLs")
    saved = await download(urls)
    print(f"Downloaded {saved} images to images/")
    build_html()


def build_html():
    images = sorted(
        p for p in IMAGES_DIR.rglob("*") if p.suffix.lower() in IMG_EXTS
    )
    if not images:
        print("No images to build gallery from.")
        return

    img_tags = "\n".join(
        f'    <img src="images/{p.name}" loading="lazy">'
        for p in images
    )

    html = f"""<!DOCTYPE html>
<html lang="ar" dir="rtl">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Gallery</title>
  <style>
    body {{ margin: 0; background: #111; }}
    .grid {{
      display: grid;
      grid-template-columns: repeat(auto-fill, minmax(240px, 1fr));
      gap: 4px;
      padding: 4px;
    }}
    img {{
      width: 100%;
      aspect-ratio: 1;
      object-fit: cover;
      display: block;
    }}
  </style>
</head>
<body>
  <div class="grid">
{img_tags}
  </div>
</body>
</html>
"""
    out = HERE / "index.html"
    out.write_text(html, encoding="utf-8")
    print(f"Built index.html with {len(images)} images")


if __name__ == "__main__":
    query = sys.argv[1] if len(sys.argv) > 1 else "تم تعبئة الكرش"
    asyncio.run(main(query))
